File: python/pro-con-python/ch2/4/UnionFindTree.py
'''
Created on 2016/12/03

@author: kazuyoshi
'''

class UFT(object):
    '''
    classdocs
    '''
    par = list()
    rank = list()

    def __init__(self, n):
        '''
        Constructor
        '''
        self.par = [i for i in range(n)]
        self.rank = [0 for i in range(n)]

    
    def find(self, x):
        if self.par[x] == x:
            return x
        else:
            self.par[x] = self.find(self.par[x])
            return self.par[x]
    
    def unite(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return
        elif self.rank[x] < self.rank[y]:
            self.par[x] = y
        else:
            self.par[y] = x
            if self.rank[x] == self.rank[y]:
                self.rank[x] += 1
        return
    
    def same(self, x, y):
        return self.find(x) == self.find(y)
    

File: c++/pro-con/ch2/2.4/UnionFindTree.cpp
#include <cstdio>
#include <algorithm>
#include <queue>

using namespace std;

const int MAX_N = 20000;

int par[MAX_N];
int Rank[MAX_N];

void init(int n){
  for(int i=0; i<n; i++){
    par[i]=i;
    Rank[i]=0;
  }
}

int find(int x){
  if (par[x] == x){
    return x;
  } else {
    return par[x] = find(par[x]);
  }
}
void unite(int x, int y) {
  x = find(x);
  y = find(y);

  if (x == y) return;

  if (Rank[x] < Rank[y]){
    par[x] = y;
  } else {
    par[y] = x;
    if (Rank[x] == Rank[y]) Rank[x]++;
  }
}

bool same(int x, int y){
  return find(x) == find(y);
}

